### DISCUSSION TITLE
correct code question - how to use plugin class file correctly in plugin?

### DISCUSSION THREAD

**Discussion by Jimmi08** [e107inc/e107#5180] [Q&A Plugins]
Hi, 
I looked at core parts and plugins (f.e. vstore)  and there are too many different solutions.
I would like to stick with one and use it in all my plugins. 

What I need to know:
- I have a custom plugin 
- plugin has  file with plugin class (imagine news handler, or forum_class or vstore class) It is class/object with static and other methods.
- plugin stuff can be displayed as single page, menu on other pages or menu on plugin single pages (before or after single page)

In the old way, you would include/ require this file.   
 
vstore is using in single page
/** @var vstore $vstore */
```
$vstore = e107::getSingleton('vstore', e_PLUGIN.'vstore/vstore.class.php');
$vstore->init();
```

but menus use:
```
require_once(e_PLUGIN . 'vstore/vstore.class.php');
$vst = new vstore;
```
so if the menu is on vstore page, there are 2 vstore classes?

if I use getSingleton() in e_module, should it be available everywhere and with e107::getSingleton('vstore')->anymethod?

The forum plugin is using
```
require_once(e_PLUGIN . 'forum/forum_class.php');
/** @var e107forum $forum */
$forum = new e107forum;

new forum_front($forum);
```

and in menu
```
include_once(e_PLUGIN.'forum/forum_class.php');
$this->forumObj = new e107forum;
```
Could somebody explain this to me?   I vaguely understand the concept (e107 is using Singleton pattern, I know), but what is new code and what is legacy stuff?  Why menus are not using  e107::getSingleton() too?   
And why in some cases is e107::getObject() used and not  e107::getSingleton()?

Thanks

**Comment by Vodhin** [comment]
I'm sure you know that `include` and `require/require_once` simply adds the text of the included file to the to the text of the file being loaded, which saves you from writing code more than once.

Use e_module.php to load your required file. It now is available for any page.
`require_once(e_PLUGIN . 'vstore/vstore.class.php');`

on your plugin page create a new instance for your class. 
`$vst = new vstore;`

on your menu create a new instance for your class.
`$vst = new vstore;`

 This doesn't load the class script twice. If you want to be nice, use `unset($vst);` at the end of your menu and page to free up server memory.

**Reply by Jimmi08** [reply]
Thank you.  I understand. But this is way how it was used in the past. 

Now there getSingleton()...   to be sure that class is called only one time.  

Just why  - if the store is only one advanced example - it can't be coded similarly- the best, recommended - way to be example for other plugins.
